### PR TITLE
Unify API for `mida_trainer` and `multi_domain_trainer`, and improve `mida_trainer` test

### LINKED ISSUE #466: [Refactor] Unify API for `mida_trainer` and `multi_domain_trainer`
`kale.pipeline.mida_trainer` is a multi-domain trainer and should be integrated into the `kale.pipeline.multi_domain_trainer`  API alongside other multi-domain trainers

### PR DESCRIPTION
**Pull request by shuo-zhou**
Fixes #466.

### Description
Move the classes and functions under `kale.pipeline.mida_trainer` to `kale.pipeline.multi_domain_trainer`. Refactor the tests related to MidaTrainer to improve speed and coverage.

Remove `mida_trainer` API from documentation.

Refactor variable name `factors` as `group_labels`.

Add `_group_label_encoder` in `MIDATrainer` to encode `group_labels` as a one-hot matrix, to support one-dimensional input of `group_labels`.

Add `LinearDiscriminantAnalysis` to `MIDATrainer` classifiers.

### Status
**In progress**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] In-line docstrings updated.
- [x] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: tests/pipeline/test_mida_trainer.py [hunk 1]
```diff
@@ -47,7 +54,7 @@ def toy_data():
 def test_mida_trainer_params_consistency(toy_data, transformer, param_grid):
     x, y, domains, factors = toy_data
 
-    param_grid = clone(param_grid, safe=False)
```
Comment (zaRizk7) [inline tests/pipeline/test_mida_trainer.py:50]: `clone` works the same way as `deepcopy` if `safe=False`

File: tests/pipeline/test_mida_trainer.py [hunk 2]
```diff
@@ -209,16 +216,19 @@ def test_mida_trainer_fit_and_methods(toy_data):
...
-    trainer.fit(x, factors=factors, groups=domains)
+    trainer.fit(x, group_labels=domains)
 
     # test score_sample
     score = trainer.score_samples(x)
     testing.assert_array_equal((len(x),), score.shape)
 
     # test transform
-    transform = trainer.transform(x, factors=factors)
+    transform = trainer.transform(x)
```
Comment (zaRizk7) [inline tests/pipeline/test_mida_trainer.py:231]: `transform` uses `factors` or `group_labels` too as it needs it when `MIDA` has `augment` set

File: kale/pipeline/multi_domain_adapter.py [hunk 3]
```diff
@@ -450,3 +476,947 @@ def decision_function(self, x):
...
+            self: The fitted trainer.
+        """
+
+        estimator = self.estimator
+        transformer = self.transformer
+        mida = MIDA() if self.use_mida else None
+        scorers, refit_metric = self._get_scorers()
+        if (1 in group_labels.shape and group_labels.ndim == 2) or (group_labels.ndim == 1):
+            params["groups"] = group_labels.copy()
+            self.group_labels_encoder = OneHotEncoder(sparse_output=False)
```
Comment (zaRizk7) [inline kale/pipeline/multi_domain_adapter.py:773]: 1. Suggest to modify `group_labels_encoder` to `_group_labels_encoder` and add a property to get the group names. The users don't need to inspect it directly.
2. We can use `LabelBinarizer` instead of `OneHotEncoder` to avoid reshaping `groups` like done on the old `MIDA` implementation.
3. `MIDA` intentionally uses `OneHotEncoder` to ignore `-1` labels, by default it is preferable to use `LabelBinarizer` instead.

Example for (1) assuming `LabelBinarizer` is used:

```py
@property
def groups_(self):
    return self._group_labels_encoder.classes_
```

File: kale/embed/factorization.py [hunk 4]
```diff
@@ -721,13 +721,13 @@ def _make_objective_kernel(self, k_x, y, factors):
         return k_x
 
     @_fit_context(prefer_skip_nested_validation=True)
-    def fit(self, x, y=None, factors=None, **fit_params):
+    def fit(self, x, y=None, group_labels=None, **fit_params):
```
[the remark this answers is not in the document]
Comment (shuo-zhou) [reply kale/embed/factorization.py:724]: `group_labels` is at least better than `factors`. We can discuss with other members for the naming of this variable in the next group meeting.

File: tests/pipeline/test_mida_trainer.py (hunk 1 above)
Comment (shuo-zhou) [reply tests/pipeline/test_mida_trainer.py:50]: There was a warning using `clone()` when I run it locally.

File: kale/evaluate/cross_validation.py [hunk 5]
```diff
@@ -65,8 +65,8 @@ def _fit_and_score(
         y (array-like): Target variable for supervised learning [num_samples] or [num_samples, num_targets].
         transformer (sklearn.base.BaseEstimator, optional): An unsupervised transformer implementing fit and transform methods applied before domain adaptation.
         domain_adapter (sklearn.base.BaseEstimator, optional): A domain adapter implementing fit and transform methods.
-        factors (array-like): The factors for adaptation with shape (num_samples, num_factors). Please preprocess the factors before domain adaptation
-                            (e.g. one-hot encode domain, gender, or standardize age).
+        group_labels (array-like): Categorical variables representing domain or grouping factors with shape
+            (num_samples, num_factors).
```
[the remark this answers is not in the document]
Comment (shuo-zhou) [reply kale/evaluate/cross_validation.py:69]: We may extend support to continuous values in the future, but currently only discrete values (categorical variables ) are supported.

**Review by zaRizk7 — APPROVED** [review]
LGTM, although for better extendability we may need to make `MIDATrainer` to be usable for multiple domain adaptation in the future to make it a more flexible to use with different ones like `TCA` or `SCA`.
